Fix part of #5134: Increased coverage to 100% for question_editor

ujjwalgulecha opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR Fixes #5134. It is to increase coverage to 100% for question_editor.

Few things. There are 3 lines of code which are not covered and that is why the coverage is not 100%. The 3 lines of code are not covered as they are never reached. 

More explanation:
[L112](https://github.com/oppia/oppia/blob/develop/core/controllers/question_editor.py#L112) will never be hit if a question does not exist. This is because an exception is raised before itself due to the [decoration](https://github.com/oppia/oppia/blob/develop/core/controllers/question_editor.py#L101) for this method, specifically [L1599](https://github.com/oppia/oppia/blob/27e7b37cf98315425881592c1da810934e0240bf/core/domain/acl_decorators.py#L1599). 

Similar reason as above for [L136](https://github.com/oppia/oppia/blob/develop/core/controllers/question_editor.py#L136)

[L141](https://github.com/oppia/oppia/blob/develop/core/controllers/question_editor.py#L141) will also never be hit because of the above reason. 

<img width="1280" alt="screen shot 2019-02-06 at 8 48 25 pm" src="https://user-images.githubusercontent.com/5632068/52391119-9231d380-2a50-11e9-965b-a9eee9e11042.png">

Do we need to create an issue for that so that those can be fixed?
## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

ujjwalgulecha committed:
Fix part of #5134: Increased coverage to 100% for question_editor

ankita240796 commented on core/controllers/question_editor_test.py:
```diff
@@ -229,6 +291,14 @@ def setUp(self):
             self.question_id, self.skill_id)
 
     def test_get(self):
+        with self.swap(constants, 'ENABLE_NEW_STRUCTURE_EDITORS', False):
```
- The function names in the test don't indicate the exact behaviour which is being tested. I see that it is throughout the file apart from your changes. Can you please try renaming all the functions to descriptive names which explain the purpose of test.

- Try testing one behaviour per test. If you are unable to name a test which indicates what behaviour it is testing, split it into multiple tests and ensure that one behaviour is tested in a single function.

For instance, you can split this test as follows:

- `test_get_with_new_structures_disabled_returns_404_status`
- `test_get_with_new_structures_enabled_does_not_allow_access_to_non_admin`
- `test_get_with_new_structures_enabled_allows_access_to_admin`
- `test_get_with_new_structures_enabled_allows_access_to_topic_manager`
- `test_get_with_new_structures_enabled_allows_access_to_question_creator`

ujjwalgulecha replied:
I have changed all the tests in the class in format stated.

ankita240796 reviewed (changes requested):
Hi @ujjwalgulecha, Thanks for the PR! I took an initial pass and there are few style related issues.

I am looking into the issue described by you in PR description and will update you shortly.

Also, in general try avoiding force push as it overwrites the history and makes it difficult to review.

Thanks!

ankita240796 commented:
Also can you please add a screenshot of the coverage report in the PR description, Thanks!

ujjwalgulecha committed:
refactored all tests in the class

ujjwalgulecha commented:
@ankita240796 made the changes. I have refactored the entire class. Also added a screenshot in the PR. Sorry for the force push, happened again! Will avoid it in the future!

aks681 commented:
@ujjwalgulecha Yeah, you can remove line 141 as the check above is directly related to that, but for the other 2 lines, the actual check in the decorator is for question_rights, right? Although technically question_rights and questions do go hand in hand, I'd suggest not removing those 2 checks, since it seems like a good second check to have. What do you think?

aks681 reviewed (commented):
From a functionality perspective, it looks good! Only one comment (see above). 
Also, a suggesion. The `ENABLE_NEW_STRUCTURE_EDITORS` flag is now flipped on develop. So, once merged with develop, you can remove all the swap statements that switches it to true.

ankita240796 commented on core/controllers/question_editor_test.py:
```diff
@@ -291,16 +403,47 @@ def test_get(self):
+                expected_status_int=404)
+            self.logout()
+
+    def test_delete_with_admin_email_deletes_question(self):
```
For consistent naming style: `test_delete_with_admin_email_allows_question_deletion`.

ujjwalgulecha replied:
👍

ankita240796 reviewed (changes requested):
Looks great @ujjwalgulecha. Thanks a lot!
Just a minor nit, PTAL.

Also, in general reply to each review comment once you have addressed it.

Thanks!

vinitamurthi reviewed (approved):
This looks good to me, thanks for making the tests so thorough!

ujjwalgulecha committed:
Merge branch 'develop' of github.com:oppia/oppia into question-editor-fullcoverage

ujjwalgulecha commented:
> @ujjwalgulecha Yeah, you can remove line 141 as the check above is directly related to that, but for the other 2 lines, the actual check in the decorator is for question_rights, right? Although technically question_rights and questions do go hand in hand, I'd suggest not removing those 2 checks, since it seems like a good second check to have. What do you think?

Yes I have removed it now. Yeah I understand that it would be better to keep it as at this point question_rights and questions do go hand in hand but later we can't say. However, then it would prevent us from reaching the desired 100% code coverage. But at this point I am leaning towards keeping the checks

ujjwalgulecha commented:
> This looks good to me, thanks for making the tests so thorough!

👍

ujjwalgulecha committed:
minor refactor

ujjwalgulecha commented:
@ankita240796 @vinitamurthi @aks681 can this be shipped if good? Thanks!

ankita240796 reviewed (approved):
LGTM, Thanks @ujjwalgulecha!

@aks681 can you please take a final pass, Thanks!

aks681 reviewed (approved):
@ujjwalgulecha Yup, sounds good, thanks!
